docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+ - Many types have some 'unused' bit patterns that can be used to embed the tag bits.
+
+In the above example, a `U8` defines 8 bits but, if stored in a 64-bit buffer, leaves 56 unused.
+The `Float64` does not have any unused bits but if the 11 exponent bits and the signalling bit are all 1, the value is a signalling not-a-number (NaN) encoding and the remaining 52 bits are unused.
+The `Foo*`, because it points to a Verona object, has zero in the low three bits and, on a typical 64-bit system with a 48-bit address space has zeroes in the top 16 bits and so needs only 45 bits to encode all valid values (on a system with a 56-bit address space, it still requires only 52 bits to represent any userspace address and so can fit in the NaN space).
+
+It is therefore possible to fully encode this example in a single 64-bit word on any common 64-bit system.
+For example, in userspace on a little-endian 64-bit system with a 48-bit virtual address space that uses integers as pointers:
+
+ - If any of bits 50-62 inclusive is non-zero, then the entire value is a `Float64`.
+ - Otherwise, if the value of bit 0 is 1, then bits 8-15 inclusive contain a `U8` value.
+ - Otherwise, bits 0-47 contain should be zero extended to give a 64-bit address.
+
+This form of encoding is one of the motivations for supporting storage-only types such as `U42`.
+Any type that does not use the full width of a machine word is amenable to dense packing in discriminated unions.
+This is a sufficiently common pattern in systems code that Verona has chosen to bake it into the language.
+
+The rules for determining the correct encoding for a union type begins by identifying the number of payload bits and the number of states required for the type discriminator:
+
+ - For singleton types, there are two possible answers:
+   - Each type requires one state in the discriminator but no bits to encode the value.
+   - Each type requires one pointer's worth of state to encode the value but no bits in the discriminator.
+   We choose the former, to maximise our ability to use the CPU's branch-on-zero instructions.
+ - Each integer requires one state in the discriminator and as many bits as its width defines for the payload.
+   For example, in the union `U8 | U16`, each requires one state in the discriminator (giving a 1-bit discriminator to represent both state) and 8 and 16 bits of payload, respectively.
+ - Floating point types follow the same rules as integers, with special handling if a floating point type is the type with the largest payload in a union type, described below.
+ - Pointers to object types (including external reference types) require as many bits as are required for an unambiguous pointer; however, all pointer types require only a single discriminator state because all pointees can be discriminated by inspecting their descriptor.
```

Comment:
I originally missed the implication of "all pointer types require only a single discriminator state"; perhaps "any number of pointer types in a union requires only a single discriminator state"?